File: C++/ReverseLinkedListII.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;

struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode *reverseBetween(ListNode *head, int m, int n) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		if(!head)
			return head;
		ListNode* start = head;
		ListNode* pre = NULL;
		ListNode* end = head;
		m--;
		n--;
		while(m--)
		{
			pre = start;
			start = start->next;
		}
		while(n--)
			end = end->next;
		while(start != end)
		{
			ListNode* next1 = start->next;
			ListNode* next2 = end->next;
			if(pre)
				pre->next = next1;
			end->next = start;
			start->next = next2;
			start = next1;
		}
		if(!pre)
			head = end;
		return head;


	}
};

File: JAVA/ReverseLinkedListII.java
import java.awt.List;

class ListNode {
	int val;
	ListNode next;

	ListNode(int x) {
		val = x;
		next = null;
	}
}

class Solution {
	public ListNode reverseBetween(ListNode head, int m, int n) {
		// Start typing your Java solution below
		// DO NOT write main() function
		if(head == null)
			return null;
		int count = 1;
		ListNode pre = null;
		ListNode p = head;
		
		while(count < m)
		{
			count++;
			pre = p;
			p = p.next;
		}
		ListNode tail = p;
		while(count < n)
		{
			
			ListNode next = tail.next;
			if(pre == null)
			{
				p = head;
				head = next;	
			}
			else
			{
				p = pre.next;
				pre.next = next;
			}
			tail.next = next.next;
			next.next = p;
			count++;
		}
		
		return head;
	}
}

public class ReverseLinkedListII {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		ListNode p1 = new ListNode(1);
		ListNode p2 = new ListNode(2);
		ListNode p3 = new ListNode(3);
		ListNode p4 = new ListNode(4);
		ListNode p5 = new ListNode(5);
		p1.next = p2;
		p2.next = p3;
		p3.next = p4;
		p4.next = p5;
		Solution sol = new Solution();
		sol.reverseBetween(p1, 3, 4);
	}

}
